crates/sdl-encoder/src/input_object_def.rs

```diff
@@ -0,0 +1,219 @@
+use crate::Field;
+use std::fmt::{self, Display};
+
+/// Input objects are composite types used as inputs into queries defined as a list of named input values..
+///
+/// InputObjectTypeDefinition
+///     Description<sub>opt</sub> **input** Name Directives<sub>\[Const\] opt</sub> FieldsDefinition<sub>opt</sub>
+///
+/// Detailed documentation can be found in [GraphQL spec](https://spec.graphql.org/draft/#sec-Input-Objects).
+///
+/// **Note**: At the moment InputObjectTypeDefinition differs slightly from the
+/// spec. Instead of accepting InputValues as `field` parameter, we accept
+/// Fields.
+///
+/// ### Example
+/// ```rust
+/// use sdl_encoder::{FieldValue, Field, InputObjectDef};
+/// use indoc::indoc;
+///
+/// let ty_1 = FieldValue::Type {
+///     ty: "DanglerPoleToys".to_string(),
+/// };
+///
+/// let ty_2 = FieldValue::List { ty: Box::new(ty_1) };
+/// let mut field = Field::new("toys".to_string(), ty_2);
+/// field.default(Some("\"Cat Dangler Pole Bird\"".to_string()));
+/// let ty_3 = FieldValue::Type {
+///     ty: "FavouriteSpots".to_string(),
+/// };
+/// let mut field_2 = Field::new("playSpot".to_string(), ty_3);
+/// field_2.description(Some("Best playime spots, e.g. tree, bed.".to_string()));
+///
+/// let mut input_def = InputObjectDef::new("PlayTime".to_string());
+/// input_def.field(field);
+/// input_def.field(field_2);
+/// input_def.description(Some("Cat playtime input".to_string()));
+///
+/// assert_eq!(
+///     input_def.to_string(),
+///     indoc! { r#"
+///         """Cat playtime input"""
+///         input PlayTime {
+///           toys: [DanglerPoleToys] = "Cat Dangler Pole Bird"
+///           """Best playime spots, e.g. tree, bed."""
+///           playSpot: FavouriteSpots
+///         }
+///     "#}
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct InputObjectDef {
+    // Name must return a String.
+    name: String,
+    // Description may return a String or null.
+    description: Option<String>,
+    // The vector of interfaces that this object implements.
+    interfaces: Vec<String>,
+    // A vector of fields
+    fields: Vec<Field>,
```

Comment:
What is the reason for accepting `Field` here instead of `InputValue` or `InputField`? I saw your comment on the PR, but I don't really understand it:

> The exception is the `__InputObject` using the prescribed by the spec `__InputValue` type. As we do not have an AST to visit to encode the schema, all types that have fields currently use `__Field` type as their input.

Since we know when a type definition is an input object by looking at the `kind` field, and every input object always has `inputFields` but not `fields`, why can't this be reflected here?

Reply:
Ok, we now have specific `InputField` encodings.